### PR TITLE
Support Mainnet / Mumbai + Refactor

### PR DESCRIPTION
**Pull request by atvanguard**
- Remove dep on matic-protocol
- clean duplicate _fillOptions
- parallelize fetch block proof
- Make some logs debug
- ⬆️@maticnetwork/meta@2.0.1

### CODE REVIEW

File: examples/utils.js [hunk 1]
```diff
@@ -1,36 +1,31 @@
...
 
 function getPrivateKey() {
-  if (!process.env.PRIVATE_KEY) {
-    throw new Error('Please set the PRIVATE_KEY env variable')
+  if (!process.env.PRIVATE_KEY || !process.env.FROM) {
+    throw new Error('Please set the PRIVATE_KEY/FROM env vars')
   }
-  return process.env.PRIVATE_KEY
+  return { privateKey: process.env.PRIVATE_KEY, from: process.env.FROM }
 }
```
Comment (arthcp) [inline examples/utils.js:29]: Can we change the function name to `getAccount` or something?

Comment (atvanguard) [reply examples/utils.js:29]: Done.

File: src/common/ContractsBase.ts [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
-import ChildERC20Artifact from 'matic-protocol/contracts-core/artifacts/ChildERC20.json'
-import ChildERC721Artifact from 'matic-protocol/contracts-core/artifacts/ChildERC721.json'
-import ChildMaticArtifact from 'matic-protocol/contracts-core/artifacts/MaticChildERC20.json'
+
+import Web3Client from './Web3Client'
+
 import { address } from '../types/Common'
 
 export default class ContractsBase {
+  static MATIC_CHILD_TOKEN: address = '0x0000000000000000000000000000000000001010'
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/common/ContractsBase.ts:8]: Nope. But we are soon gonna retire all of them.

File: src/common/ContractsBase.ts [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
+
   web3Client: Web3Client
+  network: any
 
-  constructor(web3Client: Web3Client) {
+  constructor(web3Client: Web3Client, network: any) {
     this.web3Client = web3Client
+    this.network = network
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/common/ContractsBase.ts:15]: ChildClasses are instantiating web3Client, SDKClient is - then we pass down web3Client to all child classes. Lemme know if you meant something else.

File: src/index.ts [hunk 4]
```diff
@@ -73,12 +74,32 @@ export default class Matic extends SDKClient {
   public utils: Utils
   public static MaticPOSClient = MaticPOSClient // workaround for web compatibility
 
-  constructor(options: any = {}) {
+  constructor(options: MaticClientInitializationOptions = {}) {
+    let network
+    if (options.network && options.version) {
+      network = new Network(options.network, options.version)
+      if (!network) throw new Error(`network ${options.network} - ${options.version} is not supported`)
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/index.ts:81]: Peeps using testnetV2 should use an older version of matic.js + as mentioned above we are retiring all other testnets.

File: src/index.ts [hunk 5]
```diff
@@ -73,12 +74,32 @@ export default class Matic extends SDKClient {
...
+    options = Object.assign(
+      {
+        registry: network.Main.Contracts.Registry,
+        rootChain: network.Main.Contracts.RootChainProxy,
+        depositManager: network.Main.Contracts.DepositManagerProxy,
+        withdrawManager: network.Main.Contracts.WithdrawManagerProxy,
+      },
+      options
+    )
+    options.network = network
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/index.ts:97]: Let's refactor that after we have the new design now?

File: src/libs/MerkleTree.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    return this.leaves
+  }
+
+  getLayers() {
+    return this.layers
+  }
+
+  getRoot() {
+    return this.layers[this.layers.length - 1][0]
+  }
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/libs/MerkleTree.ts:59]: Oh yeah, copied it from another repo. thanks

File: src/root/DepositManager.ts [hunk 7]
```diff
@@ -17,18 +15,18 @@ export default class DepositManager extends ContractsBase {
...
+      options.depositManager
     )
     this.registry = registry
   }
 
   async initialize() {
     const childChainAddress = (await this.registry.registry.methods.getChildChainAndStateSender().call())[0]
-    this.childChainContract = new this.web3Client.web3.eth.Contract(ChildChainArtifact.abi, childChainAddress)
+    this.childChainContract = new this.web3Client.web3.eth.Contract(this.network.abi('ChildChain'), childChainAddress)
   }
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/root/DepositManager.ts:30]: constructor can't return a promise and hence can't be awaited on.

File: src/root/RootChain.ts [hunk 8]
```diff
@@ -28,12 +32,16 @@ export default class RootChain extends ContractsBase {
...
     // binary search on all the checkpoints to find the checkpoint that contains the childBlockNumber
     let ans
-    while(start.lte(end)) {
-      if (start.eq(end)) { ans = start; break }
+    while (start.lte(end)) {
+      if (start.eq(end)) {
+        ans = start
+        break
+      }
       let mid = start.add(end).div(new BN(2))
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/root/RootChain.ts:40]: Done. Made them static variables of the class.

File: src/root/RootChain.ts [hunk 9]
```diff
@@ -28,12 +32,16 @@ export default class RootChain extends ContractsBase {
...
-      console.log({ start: start.toString(), mid: mid.toString(), end: end.toString() }) // eslint-disable-line
-      const headerBlock = await this.web3Client.call(this.rootChain.methods.headerBlocks(mid.mul(new BN(10000)).toString()))
-      // console.log('headerBlock', headerBlock)
+      logger.debug({ start: start.toString(), mid: mid.toString(), end: end.toString() }) // eslint-disable-line
+      const headerBlock = await this.web3Client.call(
+        this.rootChain.methods.headerBlocks(mid.mul(new BN(10000)).toString())
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/root/RootChain.ts:43]: Done.

File: src/root/WithdrawManager.ts [hunk 10]
```diff
@@ -133,9 +122,12 @@ export default class WithdrawManager extends ContractsBase {
...
   async startExitForMintableBurntToken(burnTxHash, predicate: address, options?) {
     const { payload, mint } = await this._buildPayloadAndFindMintTransaction(burnTxHash)
-    const _predicate = new this.web3Client.parentWeb3.eth.Contract(MintableERC721PredicateArtifact.abi, predicate)
+    const _predicate = new this.web3Client.parentWeb3.eth.Contract(
+      this.network.abi('MintableERC721Predicate'),
+      predicate
+    )
     const txObject = _predicate.methods.startExitForMintableBurntToken(payload, mint)
-    const _options = await this._fillOptions(options, txObject, this.web3Client.parentWeb3)
+    const _options = await this.web3Client.fillOptions(txObject, true /* onRootChain */, options)
```
[the remark this answers is not in the document]
Comment (atvanguard) [reply src/root/WithdrawManager.ts:130]: 😛 Discussed offline.

File: src/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (arthcp) [reply src/index.ts:97]: Does this PR break `MaticPOSClient`? Thing is dapps want to start integration as soon as possible. And matic.js surface can be same for ERC20, ether even after new design.

**Pull request merged**